PowerEditor/src/tools/NppShell/Installer.cpp

```diff
@@ -0,0 +1,326 @@
+
+    if (hResult == ERROR_SUCCESS)
+    {
+        DWORD cbData = 0;
+        hResult = RegGetValue(hkey, nullptr, valueName.c_str(), RRF_RT_REG_SZ, nullptr, nullptr, &cbData);
+
+        if (hResult == ERROR_SUCCESS)
+        {
+            vector<BYTE> buffer(cbData + 1);
+
+            cbData = (DWORD)buffer.size();
+            hResult = RegGetValue(hkey, nullptr, valueName.c_str(), RRF_RT_REG_SZ, nullptr, &buffer[0], &cbData);
+
+            wstring buildNumberString(reinterpret_cast<wchar_t*>(&buffer[0]));
+
+            int buildNumber = stoi(buildNumberString);
+
+            result = buildNumber >= FirstWindows11BuildNumber;
+        }
+    }
+
+    RegCloseKey(hkey);
+
+    return result;
+}
+
+wstring GetCLSIDString()
+{
+    auto uuid = __uuidof(NppShell::CommandHandlers::EditWithNppExplorerCommandHandler);
+
+    LPOLESTR guidString = 0;
+    HRESULT result = StringFromCLSID(uuid, &guidString);
+
+    if (FAILED(result))
+    {
+        throw "Failed to parse GUID from command handler";
+    }
+
+    wstring guid(guidString);
+
+    return guid;
+}
+
+LRESULT RemoveRegistryKeyIfFound(HKEY hive, wstring keyName)
+{
+    HKEY hkey;
+    LRESULT lResult = RegOpenKeyEx(hive, keyName.c_str(), 0, KEY_READ, &hkey);
+
+    if (lResult == ERROR_FILE_NOT_FOUND)
+    {
+        // Does not exist, so nothing to remove
+        return ERROR_SUCCESS;
+    }
+
+    if (lResult != ERROR_SUCCESS)
+    {
+        return lResult;
+    }
+
+    RegCloseKey(hkey);
+
+    return RegDeleteTree(hive, keyName.c_str());
+}
+
+LRESULT CreateRegistryKey(const HKEY hive, const wstring& key, const wstring& name, const wstring& value)
+{
+    HKEY regKey;
+    LRESULT lResult = RegCreateKeyEx(hive, key.data(), 0, NULL, REG_OPTION_NON_VOLATILE, KEY_SET_VALUE, NULL, &regKey, NULL);
+
+    if (lResult != ERROR_SUCCESS)
+    {
+        return lResult;
+    }
+
+    lResult = RegSetKeyValue(regKey, NULL, name.empty() ? NULL : name.data(), REG_SZ, value.data(), static_cast<DWORD>(value.length() * sizeof TCHAR));
+    
+    RegCloseKey(regKey);
+
+    return lResult;
+}
+
+LRESULT CleanupRegistry(wstring guid)
+{
+    int const bufferSize = MAX_PATH + GUID_STRING_SIZE;
+    WCHAR buffer[bufferSize];
+    auto arraySize = bufferSize * sizeof(WCHAR);
+
+    LRESULT result;
+
+    result = RemoveRegistryKeyIfFound(HKEY_CLASSES_ROOT, ShellExtensionKey);
+
+    if (result != ERROR_SUCCESS)
+    {
+        return result;
+    }
+
+    StringCbPrintf(buffer, arraySize, L"Notepad++_file\\shellex");
+    result = RemoveRegistryKeyIfFound(HKEY_CLASSES_ROOT, buffer);
+    if (result != ERROR_SUCCESS)
+    {
+        return result;
+    }
+
+    StringCbPrintf(buffer, arraySize, L"Software\\Classes\\CLSID\\%s", guid.c_str());
+    result = RemoveRegistryKeyIfFound(HKEY_LOCAL_MACHINE, buffer);
+    if (result != ERROR_SUCCESS)
+    {
+        return result;
+    }
+
+    return ERROR_SUCCESS;
+}
+
+LRESULT CleanupHack()
```

Comment:
We *need* to only delete this when we know for sure it's Notepad++. I see a check, but then it isn't used to guard the call to RegDeleteKey.

Reply:
Indeed you are right - I have fixed it now.
Thanks 🙂